### PR TITLE
Detect gettext tools

### PR DESCRIPTION
**Pull request by alvarez86**
I have implemented two ideas here:

1. Detect the necessary gettext tools at configuration time and warn the user if they are not found

2. Check if the tools are available at install time, and again warn the user if we can't find them.

This is meant for #821, as always, ideas and comments are welcome.

### CODE REVIEW

File: mesonbuild/modules/i18n.py [hunk 1]
```diff
@@ -12,11 +12,20 @@
...
 class I18nModule:
 
+    def has_gettext_tools(self):
+       have_all_tools = True
+       gettext_programs = ['xgettext', 'msgfmt', 'msgmerge']
+       for prog in gettext_programs:
+           dep = dependencies.ExternalProgram(prog)
+           if not dep.found():
+               have_all_tools =False
+       return have_all_tools
```
Comment (TingPing) [inline mesonbuild/modules/i18n.py]: Might as well cache this value. Also whitespace typo ^

File: mesonbuild/modules/i18n.py [hunk 2]
```diff
@@ -27,6 +36,9 @@ def gettext(self, state, args, kwargs):
         datadirs = mesonlib.stringlistify(kwargs.get('data_dirs', []))
         extra_args = mesonlib.stringlistify(kwargs.get('args', []))
 
+        if not self.has_gettext_tools():
+            mlog.warning("Some gettext tools were not found. Installation of languages won't be available until all tools are installed")
```
Comment (TingPing) [inline mesonbuild/modules/i18n.py]: Probably should only print this warning once similar to other warnings in the gnome module.

File: mesonbuild/scripts/gettext.py [hunk 3]
```diff
@@ -40,17 +41,30 @@ def run_potgen(src_sub, pkgname, datadirs, args):
         child_env['GETTEXTDATADIRS'] = datadirs
 
     ofile = os.path.join(src_sub, pkgname + '.pot')
+    msgfmt =dependecies.ExternalProgram('xgettext', silent=True)
```
Comment (TingPing) [inline mesonbuild/scripts/gettext.py]: Whitespace.

File: mesonbuild/modules/i18n.py (hunk 1 above)
Comment (alvarez86) [inline mesonbuild/modules/i18n.py]: Forgive my ignorance, but by caching you mean making it a global, or is it something else?

File: mesonbuild/modules/i18n.py (hunk 2 above)
Comment (alvarez86) [inline mesonbuild/modules/i18n.py]: I can make a logic similar to the gnome module. Trouble is, I still don't understand the situation where the warning would be printed twice. Could you give me an example to so I can work faster?

Comment (alvarez86) [inline mesonbuild/modules/i18n.py]: Fair enough, I will get to it, but it won't be done before monday. I was only thinking about the code structure, not the meson.build input, thanks.

Comment (TingPing) [inline mesonbuild/modules/i18n.py]: Seems good otherwise.

**Comment by alvarez86** [comment]
Completed the new implementation.

**Comment by jpakkane** [comment]
:crossed_swords: conflicts.

There is a fair bit of complexity here with all the different programs. Is this an actual need, that is, are there environments where a subset of Gettext tools are installed? If no, could we just require them all to be installed and refuse to work otherwise?

**Comment by alvarez86** [comment]
@jpakkane, by complexity you mean the configuration part (first 2 commits) or the installation part (3rd commit)?
If you want, I can try to change the 3rd commit to something else, if it is about the warning
at configuration time, how do you think we could it simpler?

As for the environment with only a subset of gettext, I believe it is possible, but very unlikely to happen.

**Comment by jpakkane** [comment]
A simpler version of this was merged to trunk so closing. Thanks.

**Pull request closed without merging**